src/Common/src/Microsoft/Win32/SafeHandles/GssSafeHandles.cs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Wrapper around an output gss_buffer_desc*
+    /// </summary>
+    internal sealed class SafeGssBufferHandle : SafeHandle
+    {
+        private SafeGssBufferHandle() : base(IntPtr.Zero, true)
+        {
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        internal void Copy(byte[] destination, int offset)
+        {
+            Debug.Assert(destination != null, "destination buffer cannot be null");
+            Debug.Assert(offset <= destination.Length , "offset in the destination buffer must be valid");
+            if (destination.Length > 0)
+            {
+                Interop.NetSecurityNative.CopyBuffer(this, destination, capacity: destination.Length, offset: offset);
+            }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ReleaseBuffer(out minorStatus, handle);
+            SetHandle(IntPtr.Zero);
+            return status == Interop.NetSecurityNative.Status.GSS_S_COMPLETE;
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_name_t_desc*
+    /// </summary>
+    internal sealed class SafeGssNameHandle : SafeHandle
+    {
+        public static SafeGssNameHandle Create(string name, bool isUser)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name), "Invalid name passed to SafeGssNameHandle create");
```

Comment:
Where does this name originally come from?  Is higher-level code responsible for validating it such that even with bad input we'd never expect this assert to fire?